superset-frontend/src/middleware/asyncEvent.ts

```diff
@@ -0,0 +1,188 @@
+        endpoint: url,
+      });
+
+      return json.result;
+    };
+
+    const fetchCachedData = async (
+      asyncEvent: AsyncEvent,
+      componentId: number,
+    ): Promise<CachedDataResponse> => {
+      let status = 'success';
+      let data;
+      try {
+        const { json } = await SupersetClient.get({
+          endpoint: asyncEvent.result_url,
+        });
+        data = 'result' in json ? json.result[0] : json;
+      } catch (response) {
+        status = 'error';
+        data = await getClientErrorObject(response);
+      }
+
+      return { componentId, status, data };
+    };
+
+    const setLastId = (asyncEvent: AsyncEvent) => {
+      lastReceivedEventId = asyncEvent.id;
+      try {
+        localStorage.setItem(LOCALSTORAGE_KEY, lastReceivedEventId as string);
+      } catch (err) {
+        console.warn('Error saving event ID to localStorage', err);
+      }
+    };
+
+    const processEvents = async () => {
+      const state = store.getState();
+      const queuedComponents = getPendingComponents(state);
+      let events: AsyncEvent[] = [];
+      if (queuedComponents && queuedComponents.length) {
+        try {
+          events = await fetchEvents(lastReceivedEventId);
+          if (events && events.length) {
+            const componentsByJobId = queuedComponents.reduce((acc, item) => {
+              acc[item.asyncJobId] = item;
+              return acc;
+            }, {});
+            const fetchDataEvents: Promise<CachedDataResponse>[] = [];
+            events.forEach((asyncEvent: AsyncEvent) => {
+              const component = componentsByJobId[asyncEvent.job_id];
+              if (!component) {
+                console.warn(
+                  'component not found for job_id',
+                  asyncEvent.job_id,
+                );
+                return false;
+              }
+              const componentId = component.id;
+              switch (asyncEvent.status) {
+                case JOB_STATUS.DONE:
+                  fetchDataEvents.push(
+                    fetchCachedData(asyncEvent, componentId),
+                  );
+                  break;
+                case JOB_STATUS.ERROR:
+                  store.dispatch(
+                    errorAction(componentId, parseErrorJson(asyncEvent)),
+                  );
+                  break;
+                default:
+                  console.warn('received event with status', asyncEvent.status);
+              }
+
+              return setLastId(asyncEvent);
+            });
+
+            const fetchResults = await Promise.all(fetchDataEvents);
+            fetchResults.forEach(result => {
+              if (result.status === 'success') {
+                store.dispatch(successAction(result.componentId, result.data));
+              } else {
+                store.dispatch(errorAction(result.componentId, result.data));
+              }
+            });
+          }
+        } catch (err) {
+          console.warn(err);
```

Comment:
Is there a way to fallback to regular API requests when async events failed to initialize for some reason?

We should probably also stop sending further polling requests for certain errors. E.g. this access denied error:

<img src="https://user-images.githubusercontent.com/335541/101556680-431c9880-3970-11eb-8414-2e7e5e9e4c22.png" width="400" />

Reply:
Do you have any log data on what error is being raised here?

Reply:
I think it's because I didn't run `superset init` to update the user access table.